modules/core/src/main/java/org/apache/ignite/lang/IgniteLogger.java

```diff
@@ -55,15 +56,23 @@ protected IgniteLogger(@NotNull Class<?> cls) {
      * @param params Parameters.
      */
     public void info(String msg, Object... params) {
-        log.log(INFO, msg, params);
+        if (log.isLoggable(INFO)) {
+            IgniteBiTuple<String, Throwable> readyParams = format(msg, params);
+
+            log.log(INFO, readyParams.get1(), readyParams.get1());
+        }
```

Comment:
Actually, throwable is not a part of the message template argument.
So, assuming throwable in params looks like a contract violation.

Using vararg for passing anything else other than template parameters looks error-prone and
generally, it is bad practice to pass objects of a different kind as vararg.

Reply:
It might worth having "format(String, Object...)" method in IgniteLogger itself.

IMHO, precalculating message in the particular case looks ok if you already going to log a Throwable.
Because the performance becomes much less important once you already have thread stacktrace materialized.

Reply:
I slightly changed javadoc in order to allow to pass a Throwble in the latest parameter.
I am not sure that API which require making format by hand would be more pretty... Although, I want to see an opinion from anyone else.

Reply:
I made a small investigation around different logging interfaces  (jul, log4j, System.Logger) and it looks like only SLF4J using the doubtful approach with Throwable as the last param in varargs.

So, I would prefer something like:
`error(Supplier<String> msgSupplier, Throwable th)`
and calling format manually on the user side. It must be a known API for users of modern JUL, log4j, System.Logger versions.

But also we should provide shorthand versions for methods, which don't want to provide Throwable at all:
`info(String format, Object... params)`
and etc.

WDYT @vldpyatkov

Reply:
Two against one.
Changed.